exp/api/v1beta1/azuremachinepool_default.go

```diff
@@ -76,25 +76,26 @@ func (amp *AzureMachinePool) SetDefaultSSHPublicKey() error {
 func (amp *AzureMachinePool) SetIdentityDefaults(subscriptionID string) {
 	// Ensure the deprecated fields and new fields are not populated simultaneously
 	if amp.Spec.RoleAssignmentName != "" && amp.Spec.SystemAssignedIdentityRole != nil && amp.Spec.SystemAssignedIdentityRole.Name != "" {
-		// Both the deprecated and the new fields are both set, return without changes
-		// and reject the request in the validating webhook which runs later.
-		return
```

Comment:
Apologies if my issue description wasn't clear. `azuremachinepool_default.go` does not need any code changes. Currently the defaulter is missing unit tests which are defined in `azuremachinepool_default_test.go` here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_default_test.go

In this case, you would add a unit test case to the `TestAzureMachinePool_SetIdentityDefaults` function: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_default_test.go#L50

Please let me know if you have any questions regarding this!

Reply:
Certainly, I apologize for not fully understanding your initial description of the issue. Thank you for clarifying it again that the issue is related to missing unit tests in the "azuremachinepool_default_test.go" file.

Does this look correct for **SetDiagnosticsDefaults()**? 

- I added expect value to **To(Not(BeNil()))** for each `Diagnostics.Boot`

![image](https://user-images.githubusercontent.com/83772913/232883231-60f02fe2-aa33-44c5-9eb7-0b6b7849696b.png)

I am interested in working on this issue. However, I may require some guidance and assistance in completing this task. Thankyou.

Reply:
**Obviously there will be mistakes** but, 
I tried writing according to your guidance for **SetDiagnosticsDefaults()**.
which of this two looks correct? or both are wrong?
```
bootDefault := &infrav1.BootDiagnostics{
	StorageAccountType: infrav1.ManagedDiagnosticsStorage,
}
nilBootDiagnostics := test{machinePool: &AzureMachinePool{
	Spec: AzureMachinePoolSpec{
		Template: AzureMachinePoolMachineTemplate{
			Diagnostics: &infrav1.Diagnostics{
				Boot: &infrav1.BootDiagnostics{
                                        StorageAccountType: infrav1.ManagedDiagnosticsStorage,
                                },
			},
		},
	},
}}
nilBootDiagnostics.machinePool.SetDiagnosticsDefaults()
g.Expect(nilBootDiagnostics.machinePool.Spec.Template.Diagnostics.Boot).To(Equal(bootDefault))
```
OR
```
nilBootDiagnostics := test{machinePool: &AzureMachinePool{
	Spec: AzureMachinePoolSpec{
		Template: AzureMachinePoolMachineTemplate{
			Diagnostics: &infrav1.Diagnostics{
				Boot: &infrav1.BootDiagnostics{
                                        StorageAccountType: infrav1.ManagedDiagnosticsStorage,
                                },
		        },
		},
	},
}}
nilBootDiagnostics.machinePool.SetDiagnosticsDefaults()
g.Expect(nilBootDiagnostics.machinePool.Spec.Template.Diagnostics.Boot).To(Not(BeNil()))
```
----------------------------------------------------------------------------------------------------------
I also tried for **SetIdentityDefaults()** : 
```
bothFieldsPopulatedTest := test{machinePool: &AzureMachinePool{Spec: AzureMachinePoolSpec{
	Identity:           infrav1.VMIdentitySystemAssigned,
	RoleAssignmentName: existingRoleAssignmentName,
	SystemAssignedIdentityRole: &infrav1.SystemAssignedIdentityRole{},
}}}
bothFieldsPopulatedTest.machinePool.SetIdentityDefaults(fakeSubscriptionID)
g.Expect(bothFieldsPopulatedTest.machinePool.Spec.RoleAssignmentName).To(BeEmpty())
g.Expect(bothFieldsPopulatedTest.machinePool.Spec.SystemAssignedIdentityRole.Name).To(BeEmpty())
```
----------------------------------------------------------------------------------------------------------
Also need help for **SetSpotEvictionPolicyDefaults()**.

Reply:
Thankyou,
I will be adding below given block to **TestAzureMachinePool_SetIdentityDefaults**
```
bothFieldsPopulatedTest := test{machinePool: &AzureMachinePool{Spec: AzureMachinePoolSpec{
	Identity:           infrav1.VMIdentitySystemAssigned,
	RoleAssignmentName: existingRoleAssignmentName,
	SystemAssignedIdentityRole: &infrav1.SystemAssignedIdentityRole{
		Name: existingRoleAssignmentName,
	},
}}}
bothFieldsPopulatedTest.machinePool.SetIdentityDefaults(fakeSubscriptionID)
g.Expect(bothFieldsPopulatedTest.machinePool.Spec.RoleAssignmentName).To(BeEmpty())
g.Expect(bothFieldsPopulatedTest.machinePool.Spec.SystemAssignedIdentityRole).To(BeEmpty())
```
And given below block to **TestAzureMachinePool_SetDiagnosticsDefaults**
```
bootDefault := &infrav1.BootDiagnostics{
	StorageAccountType: infrav1.ManagedDiagnosticsStorage,
}
nilBootDiagnostics := test{machinePool: &AzureMachinePool{
	Spec: AzureMachinePoolSpec{
		Template: AzureMachinePoolMachineTemplate{
			Diagnostics: &infrav1.Diagnostics{},
		},
	},
}}
nilBootDiagnostics.machinePool.SetDiagnosticsDefaults()
g.Expect(nilBootDiagnostics.machinePool.Spec.Template.Diagnostics.Boot).To(Equal(bootDefault))
```

And will be creating new test function for **SetSpotEvictionPolicyDefaults()** : 
```
func TestAzureMachinePool_SetSpotEvictionPolicyDefaults(t *testing.T) {
	g := NewWithT(t)

	type test struct {
		machinePool *AzureMachinePool
	}
        // test for SetSpotEvictionPolicyDefaults() to Ensure the default policy is being set correctly depending on the input
	defaultPolicySetCorrectly1 := test{machinePool: &AzureMachinePool{
		Spec: AzureMachinePoolSpec{
			Template: AzureMachinePoolMachineTemplate{
				SpotVMOptions: &infrav1.SpotVMOptions{
					EvictionPolicy: infrav1.SpotEvictionPolicyDeallocate,
				},
			},
		},
	}}
	defaultPolicySetCorrectly1.machinePool.SetSpotEvictionPolicyDefaults()
	g.Expect(defaultPolicySetCorrectly1.machinePool.Spec.Template.SpotVMOptions.EvictionPolicy).To(Equal(infrav1.SpotEvictionPolicyDeallocate))

	// test for SetSpotEvictionPolicyDefaults() to Ensure the default policy is being set correctly depending on the input
	defaultPolicySetCorrectly2 := test{machinePool: &AzureMachinePool{
		Spec: AzureMachinePoolSpec{
			Template: AzureMachinePoolMachineTemplate{
				SpotVMOptions: &infrav1.SpotVMOptions{},
				OSDisk: AzureOSDisk{
					DiffDiskSettings: &infrav1.DiffDiskSettings{
						Option: "Local",
					},
				},
			},
		},
	}}
	defaultPolicySetCorrectly2.machinePool.SetSpotEvictionPolicyDefaults()
	g.Expect(defaultPolicySetCorrectly2.machinePool.Spec.Template.SpotVMOptions.EvictionPolicy).To(Equal(infrav1.SpotEvictionPolicyDelete))
}
```
Still sure that there will be **Errors**.

Reply:
`TestAzureMachinePool_SetIdentityDefaults`:

That is close! I may have misspoke... Because the function is returning early, it won't make any modifications to the given `AzureMachinePoolSpec`. Thus, you should ensure that both RoleAssignmentName and SystemAssignedIdentityRole.Name are equal to existingRoleAssignmentName.

`TestAzureMachinePool_SetDiagnosticsDefaults`:

That looks perfect! Good job 😄 

`SetSpotEvictionPolicyDefaults()`:

Your second test case is correct! For the first test case, you still want SpotVMOptions.EvictionPolicy to be nil, otherwise you won't enter the first if block.


Thanks for all your hard work on this! In addition, if you're using VSCode for development, it may help you to run these tests locally. If you need help with how to setup a local test environment, please let me know!

Reply:
Thankyou for helping me out 😄
Changes in first test case of **TestAzureMachinePool_SetSpotEvictionPolicyDefaults()** : 
```
defaultPolicySetCorrectly1 := test{machinePool: &AzureMachinePool{
	Spec: AzureMachinePoolSpec{
		Template: AzureMachinePoolMachineTemplate{
			SpotVMOptions: &infrav1.SpotVMOptions{},
		},
	},
}}
defaultPolicySetCorrectly1.machinePool.SetSpotEvictionPolicyDefaults()
g.Expect(defaultPolicySetCorrectly1.machinePool.Spec.Template.SpotVMOptions.EvictionPolicy).To(BeNil())
```
If this looks correct than I can make a PR 😄 
Yes, I am using VSCode for development and need help to setup a local test environment.

Reply:
Close.. you still want to check if the eviction policy is set to `Deallocate`!

Great! If you have VSCode, make sure you have Go 1.19 and the Go extension installed. You should be able to have an option to run the test above the test function. Let me know if you see that option!

Reply:
Thanks, I have  Go v1.20.3 . Following your steps helped me to get the run test functionality in my VSCode.
![image](https://user-images.githubusercontent.com/83772913/233787086-010bd496-fe40-4498-836c-ed4d15166fe9.png)

It also helped me to write the tests perfectly and they all ran successfully. Here is the screenshot : 
![image](https://user-images.githubusercontent.com/83772913/233787140-7a90df83-0750-472c-89d1-fdf01dafdf0b.png)
After your help/guidance  I have created one pull request : https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3475

Hope this is correct one 😄 .

Reply:
Sure we can close but first, I wanted to take a moment to express my sincere gratitude for all the help you provided me in my contribution. Your expertise and guidance were instrumental in helping me write effective tests.

Your patience and willingness to answer all my questions were truly appreciated. I learned so much from you during this process, and I feel much more confident in my testing abilities because of your help.

I am grateful for the time and effort you put into helping me, and I wanted to let you know how much it means to me. I am grateful to have had the opportunity to work with you.

Appreciate your efforts for helping me sir ❤️.

Reply:
Also I am interested in learning more about the project and potentially becoming a member to contribute and learn more in the future. I am excited about the prospect of working with such a talented and knowledgeable team. If you could guide me further for becoming the member😄!